code_generation/optimizer.py

```diff
@@ -0,0 +1,292 @@
+                log.debug("Input {} not in outputs".format(input))
+        return wrong_inputs
+
+    """
+    Function used to locate the producers responsible for creating the given inputs.
+    The function return a list of producers, that have to be run before the tested producer.
+    If a needed input is not found, the function raise an Exception.
+
+    Args:
+        producer: The producer to check
+        inputs: The list of inputs to check
+    Returns:
+        producers_to_relocate: A list of producers, that have to be run before the given producer
+    """
+
+    def find_inputs(self, producer, inputs):
+        producers_to_relocate = set()
+        log.debug("Trying to find inputs {}".format(inputs))
+        for input in inputs:
+            found = False
+            for producer in self.ordering:
+                if input in producer.get_outputs(self.scope):
+                    found = True
+                    log.debug(
+                        "found {} in outputs from producer {} ({}) --> Rank {}".format(
+                            input,
+                            producer,
+                            producer.get_outputs(self.scope),
+                            self.ordering.index(producer),
+                        )
+                    )
+                    producers_to_relocate.add(producer)
+            if self.scope != "global":
+                for producer in self.global_producers:
+                    if input in producer.get_outputs("global"):
+                        found = True
+                        log.debug("found {} in global scope ..".format(input))
+            if not found:
+                log.error("{} was not found in any producer output!".format(input))
+                log.error("Please check if all needed producers are activated !")
+                raise Exception
+        return list(producers_to_relocate)
+
+    """
+    Function used to relocate a producer to a given position.
+
+    Args:
+        producer: The producer to relocate
+        old_position: The old position of the producer in the ordering
+        new_position: The new position of the producer in the ordering
+    """
+
+    def relocate_producer(self, producer, old_position, new_position):
+        log.debug(
+            "Relocating Producer {} from rank {} to rank {}".format(
+                producer, old_position, new_position
+            )
+        )
+        updated_ordering = list(range(self.size))
+        if old_position > new_position:
+            for position in updated_ordering:
+                if position <= old_position and position > new_position:
+                    updated_ordering[position] = position - 1
+                if position == new_position:
+                    updated_ordering[position] = old_position
+        if old_position < new_position:
+            for position in updated_ordering:
+                if position < old_position and position >= new_position:
```

Comment:
this is never true. probably mixed up old and new

Reply:
yeah that is true. 

This part is mainly implemented for completeness. The way the algorithm is implemented now, this case should never happen, we always want to put producers from a higher position to a lower position.